### PR TITLE
feat: support fp16 lm head gemm

### PR DESCRIPTION
**Pull request by Vinkle-hzt**
## Summary

- add an FP16 x FP16 cuBLAS GEMM with FP32 accumulation and output
- use the FP16 GEMM for both PyWrappedModel LM-head post-layer paths

### CODE REVIEW

**Review by netaddi — APPROVED** [review]

File: rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc [hunk 1]
```diff
@@ -67,4 +67,63 @@ at::Tensor cublas_gemm_bf16_bf16_fp32(const at::Tensor& input, const at::Tensor&
     return out;
 }
 
+at::Tensor cublas_gemm_fp16_fp16_fp32(const at::Tensor& input, const at::Tensor& weight) {
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70]: **[P2] 新增 FP16 GEMM 缺少聚焦数值与边界测试**

新实现手工处理 cuBLAS 矩阵布局、FP32 累加、非连续输入及零维分支，并替换 CUDA FP16 LM-head 路径，但现有直接 GEMM 测试仅覆盖 BF16，无法稳定发现 FP16 输出 dtype、转置方向、leading dimension 和空维度回归。

**建议:** 增加 CUDA 单测，与 `input.float().matmul(weight.float().t())` 对比，覆盖常规及非方阵、非连续输入、`M/N/K == 0` 和 FP32 输出类型；同时覆盖普通或 CP last-hidden 的 FP16 集成路径。

**Checklist:** [6.1] 新逻辑有聚焦单测 + 相关集成/smoke 测试；[6.1] 边界 case 覆盖（空、单元素、最大值）

**Review by LLLLKKKK — COMMENTED** [review]
## AI Code Review - PR #1389

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/1 · P3/0

**Reviewed:** commit `826b5ed38520` · 2026-09-07 17:19 UTC+8

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **新增 FP16 GEMM 缺少聚焦数值与边界测试** @ `rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70`
  - 建议：增加 CUDA 单测，与 `input.float().matmul(weight.float().t())` 对比，覆盖常规及非方阵、非连续输入、`M/N/K == 0` 和 FP32 输出类型；同时覆盖普通或 CP last-hidden 的 FP16 集成路径。

### Checklist Findings (2 fail / 64 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `新增 FP16 GEMM 缺少聚焦数值与边界测试`
  _新实现手工处理 cuBLAS 矩阵布局、FP32 累加、非连续输入及零维分支，并替换 CUDA FP16 LM-head 路径，但现有直接 GEMM 测试仅覆盖 BF16，无法稳定发现 FP16 输出 dtype、转置方向、leading dimension 和空维度回归。_
- [6.1] Tests — 边界 case 覆盖（空、单元素、最大值）  → issue `新增 FP16 GEMM 缺少聚焦数值与边界测试`
  _新实现手工处理 cuBLAS 矩阵布局、FP32 累加、非连续输入及零维分支，并替换 CUDA FP16 LM-head 路径，但现有直接 GEMM 测试仅覆盖 BF16，无法稳定发现 FP16 输出 dtype、转置方向、leading dimension 和空维度回归。_

### Strengths
- 完整校验 device、dtype、维度、同设备约束及 cuBLAS int32 上限。
- 正确处理非连续 tensor、空输出和 `K == 0`，并直接生成 FP32 logits。
- 普通与 CP last-hidden 路径保持一致，非 CUDA 平台继续使用原有 fallback。

**Review by netaddi — COMMENTED** [review]
## AI Code Review - PR #1389

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/0 · P3/0

**Reviewed:** commit `f49713e8a1f9` · 2026-09-09 15:31 UTC+8

lgtm ready to ci

### Checklist ✅ (26 items passed)

### Strengths
- 无法基于源码确认具体优点。

**Comment by wht21** [comment]
internal source has been updated, please review the changes!

File: rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc (hunk 1 above)
Comment (LLLLKKKK) [inline rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70]: **[P2] 新增 FP16 GEMM 缺少聚焦数值与边界测试**

新函数手工指定转置、leading dimension、FP32 累加及零维分支，并替换两条 CUDA FP16 LM-head 路径，但本次未增加测试。现有直接 GEMM 单测仅覆盖 BF16，无法精准发现 FP16 输出 dtype、矩阵方向、非连续输入或空维度回归。

**建议:** 增加 CUDA 单测，与 `input.float().matmul(weight.float().t())` 对比，覆盖非方阵、非连续输入、FP32 输出、`M/N/K == 0` 及 FP16 中间结果会溢出的输入；同时覆盖普通和 CP last-hidden FP16 集成路径。

**Checklist:** [6.1] 新逻辑有聚焦单测 + 相关集成/smoke 测试；[6.1] 边界 case 覆盖（空、单元素、最大值）

Comment (LLLLKKKK) [inline rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70]: **[P2] BF16 与 FP16 GEMM 实现重复，关键约束容易漂移**

FP16 函数完整复制 BF16 实现，除错误文本和输入数据类型外基本一致。后续修正设备校验、连续化、空维度、stream 或 cuBLAS 布局时需要同步两份代码，容易形成行为差异。

**建议:** 保留现有公开包装函数，将公共校验、连续化、输出分配及 `cublasGemmEx` 调用收敛到私有模板或 dtype 参数化实现。

**Checklist:** [6.1] DRY：重复非平凡逻辑被抽取或显式复用；[I] 同一功能用统一工具函数

**Review by LLLLKKKK — COMMENTED** [review]
## AI Code Review - PR #1389

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/2 · P3/0

**Reviewed:** commit `cd70a377a50c` · 2026-09-11 10:35 UTC+8

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **新增 FP16 GEMM 缺少聚焦数值与边界测试** @ `rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70`
  - 建议：增加 CUDA 单测，与 `input.float().matmul(weight.float().t())` 对比，覆盖非方阵、非连续输入、FP32 输出、`M/N/K == 0` 及 FP16 中间结果会溢出的输入；同时覆盖普通和 CP last-hidden FP16 集成路径。
- **BF16 与 FP16 GEMM 实现重复，关键约束容易漂移** @ `rtp_llm/models_py/bindings/cuda/Bf16GemmOp.cc:70`
  - 建议：保留现有公开包装函数，将公共校验、连续化、输出分配及 `cublasGemmEx` 调用收敛到私有模板或 dtype 参数化实现。

### Checklist Findings (4 fail / 77 total)

**General Principles Checklist**
- [6.1] Software Engineering — DRY：重复非平凡逻辑被抽取或显式复用  → issue `BF16 与 FP16 GEMM 实现重复，关键约束容易漂移`
  _FP16 函数完整复制 BF16 实现，除错误文本和输入数据类型外基本一致。后续修正设备校验、连续化、空维度、stream 或 cuBLAS 布局时需要同步两份代码，容易形成行为差异。_
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `新增 FP16 GEMM 缺少聚焦数值与边界测试`
  _新函数手工指定转置、leading dimension、FP32 累加及零维分支，并替换两条 CUDA FP16 LM-head 路径，但本次未增加测试。现有直接 GEMM 单测仅覆盖 BF16，无法精准发现 FP16 输出 dtype、矩阵方向、非连续输入或空维度回归。_
- [6.1] Tests — 边界 case 覆盖（空、单元素、最大值）  → issue `新增 FP16 GEMM 缺少聚焦数值与边界测试`
  _新函数手工指定转置、leading dimension、FP32 累加及零维分支，并替换两条 CUDA FP16 LM-head 路径，但本次未增加测试。现有直接 GEMM 单测仅覆盖 BF16，无法精准发现 FP16 输出 dtype、矩阵方向、非连续输入或空维度回归。_

**RTP-LLM Checklist**
- [I] 代码质量 — 同一功能用统一工具函数  → issue `BF16 与 FP16 GEMM 实现重复，关键约束容易漂移`
  _FP16 函数完整复制 BF16 实现，除错误文本和输入数据类型外基本一致。后续修正设备校验、连续化、空维度、stream 或 cuBLAS 布局时需要同步两份代码，容易形成行为差异。_

### Strengths
- 完整校验设备、dtype、维度及 cuBLAS int32 上限，并正确处理非连续张量和零维输入。
- FP16 路径直接生成 FP32 logits，避免 FP16 中间输出及后续转换。
- 普通与 CP last-hidden 路径保持一致，非 CUDA fallback 未改变。

**Comment by wht21** [comment]
internal source has been updated, please review the changes!
